src/main/java/bio/terra/service/resourcemanagement/google/GoogleProjectService.java

```diff
@@ -401,6 +418,102 @@ private CloudResourceManager cloudResourceManager() throws IOException, GeneralS
+     * @param appengine appengine client
+     * @param googleProjectId name of the google project to create the Firestore DB in
+     * @param locationId location of the Firestore DB
+     * @param timeout how long to wait for the creation operation
+     */
+    private static void enableFirestore(final Appengine appengine,
+                                        final String googleProjectId,
+                                        final String locationId,
+                                        final long timeout) throws IOException, InterruptedException {
+        logger.info("Enabling Firestore in project {} in location {}", googleProjectId, locationId);
+        // Create a request object
+        Appengine.Apps.Create createRequest = appengine.apps().create(new Application()
+            .setId(googleProjectId)
+            .setLocationId(locationId)
+            .setDatabaseType(FIRESTORE_DB_TYPE)
+        );
+
+        // Make sure that Firestore is created in the correct project
+        createRequest.getRequestHeaders().set("X-Goog-User-Project", googleProjectId);
+
+        // Execute the request
+        com.google.api.services.appengine.v1.model.Operation operation = createRequest.execute();
+
+        // Wait for the Firestore creation to finish
+        blockUntilAppengineOperationComplete(appengine, operation, googleProjectId, timeout);
+        logger.info("Firestore was enabled successfully");
+    }
+
+    /**
+     * Poll the app engine api until an operation completes. It is possible to hit quota issues
+     * here, so the timeout is set to 10 seconds.
+     *
+     * @param appengine       service instance
+     * @param operation       has an id for us to use in the check
+     * @param appId           id of the app that the operation is related to
+     * @param timeoutSeconds  how many seconds before we give up
+     * @return a completed operation
+     */
+    private static com.google.api.services.appengine.v1.model.Operation blockUntilAppengineOperationComplete(
+        Appengine appengine,
+        com.google.api.services.appengine.v1.model.Operation operation,
+        String appId,
+        long timeoutSeconds) throws IOException, InterruptedException {
+        long start = System.currentTimeMillis();
+        final long pollInterval = 10 * 1000; // 10 seconds
+        // Get
+        final String opName = operation.getName();
+        // The format returns is apps/{appId}/operations/{useful id} so we need to extract it
+        // Add a check in case they ever change the format
+        if (!opName.matches(String.format("^apps/%s/operations/" +
+            "[0-9a-fA-F]{8}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{4}\\-[0-9a-fA-F]{12}", appId))) {
+            throw new AppengineException(String.format("Operation Name does not look as expected: %s", opName));
+        }
+        final String opId = opName.replaceAll(String.format("^apps/%s/operations/", appId), "");
```

Comment:
It would be nice to avoid the code duplication here. I think you can do both the check on line 491 and the extraction of the opId by using a capture group in the regex. Also, I don't think `-` needs to be quoted since `-` is only special inside a character class. Also there's a special form for hex digits.

Not tested, but something like this should work:

```
Pattern pattern = Pattern.compile(String.format("^apps/%s/operations/(\p{XDigit}{8}-\p{XDigit}{4}-\p{XDigit}{4}-\p{XDigit}{4}-\p{XDigit}{12})", appId));
Matcher matcher = pattern.matcher(opName);
if (!matcher.find()) {
    throw ...;
}
final String opId = matcher.group(1);
```
(like the original code, this code assumes that `appId` doesn't contain any regex chars that need to be quoted)

Reply:
of interesting about the hex digit bit.  And yeah, fair point about the extraction.  I'll pull this into a separate unit-testable method